cachi2/core/package_managers/npm.py

```diff
@@ -173,6 +185,199 @@ def get_sbom_components(self) -> list[dict[str, str]]:
+
+    return vcs_url_info
+
+
+def _clone_repo_pack_archive(
+    vcs: str,
+    download_dir: RootedPath,
+) -> RootedPath:
+    """
+    Clone a repository and pack its content as tar.
+
+    :param url: URL for file download
+    :param download_dir: Output folder where dependencies will be downloaded
+    :raise FetchError: If download failed
+    """
+    info = _extract_git_info_npm(vcs)
+    download_path = RootedPath(
+        Path(
+            download_dir,
+            info["host"],  # host
+            info["namespace"],
+            info["repo"],
+            f'{info["repo"]}-external-gitcommit-{info["ref"]}.tgz',
+        )
+    )
+
+    # Create missing directories
+    directory = os.path.dirname(download_path)
+    os.makedirs(directory, exist_ok=True)
+    clone_as_tarball(info["url"], info["ref"], download_path.path)
+
+    return download_path
+
+
+def _get_npm_dependencies(
+    download_dir: RootedPath, deps_to_download: Dict[str, Dict[str, str]]
+) -> Dict[str, RootedPath]:
+    """
+    Download npm dependencies.
+
+    Receives the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param download_dir: Destination directory path where deps will be downloaded
+    :param deps_to_download: Dict of dependencies to be downloaded.
+    :return: Dictionary of dependencies with downloaded paths
+    """
+    files_to_download: dict[str, dict[str, Any]] = {}
+    download_paths = {}
+    for dependency, info in deps_to_download.items():
+        if (
+            info["url"].startswith("github:")
+            or info["url"].startswith("gitlab:")
+            or info["url"].startswith("bitbucket:")
+        ):
+            info["url"] = _update_vcs_url_with_full_hostname(info["url"])
+
+        if info["url"].startswith("git+"):
+            # Git source
+            download_paths[dependency] = _clone_repo_pack_archive(info["url"], download_dir)
+        else:
+            if "registry.npmjs.org" == urlparse(info["url"]).hostname:
+                # Tar archive from npm registry
+                archive_name = f'{dependency}-{info["version"]}.tgz'.removeprefix("@").replace(
+                    "/", "-"
+                )
+                download_paths[dependency] = download_dir.join_within_root(archive_name)
+            else:
+                # Tar archive
+                if info["integrity"]:
+                    algorithm, digest = ChecksumInfo.from_sri(info["integrity"])
+                else:
+                    raise FetchError(
+                        f"{dependency} is missing integrity checksum. It is mandatory"
+                        f"for https dependencies.",
+                        solution="Please double-check provided package-lock.json that"
+                        " your dependencies specify integrity. Try to "
+                        "rerun `npm install` on your repository.",
+                    )
```

Comment:
This is not a FetchError - a fetch error would be if cachi2 tries to download something and fails. Here, we're failing the request before we even try to download because the lockfile doesn't meet the requirements

IMO this is PackageRejected